### ISSUE TITLE
Incorrect error message when asserting an interface

### ISSUE THREAD

**Issue by LuxDie** [Microsoft/TypeScript#8653]
**TypeScript Version:** 

1.8.10

**Code**

This snippet will throw the error "neither type is assignable to each other, `property` is missing"
```ts
interface IMyInterface  {
  func(num: number): void;
  property: boolean;
}

let myObj: IMyInterface;

let otherObj = {
  func(): void {}
};

myObj = <IMyInterface>otherObj;
```
but when adding a parameter to our method the error will dissapear
```ts
interface IMyInterface  {
  func(num: number): void;
  property: boolean;
}

let myObj: IMyInterface;

let otherObj = {
  func(num): void {}
};

myObj = <IMyInterface>otherObj;
```
also fixing what the error complains about will remove the error, but I shouldn't need to add `property` when already asserting the type:
```ts
interface IMyInterface  {
  func(num: number): void;
  property: boolean;
}

let myObj: IMyInterface;

let otherObj = {
  func(num): void {},
  property: false
};

myObj = <IMyInterface>otherObj;
```   
**Expected behavior:**
No error message

**Actual behavior:** 
An error message is thrown

**Comment by RyanCavanaugh** [comment]
Currently when you write `<A>b`, we check if the type of `b` (let's call it `B`) is assignable to `A`, or vice versa.

In this case, `A` is not assignable to `B` because an extra parameter is required, and `B` is not assignable to `A` because an extra property is required.

If you accept the premise that *some assertions should be invalid* (which I think is useful, e.g. disallowing you to assert a `number` to an `HTMLElement`), then the question is why *this* example should be allowed (despite having disjoint requirements) but not others.

#5300 and #1447 tracks having some relation that would handle this more directly, but at some point it does seem correct to issue an error.

**Comment by martine** [comment]
There's also the *force harder* button available to coerce anything.  The code is extra ugly to warn you away from doing it:

```js
let x: HTMLElement = <HTMLElement><any>3;
```

**Comment by LuxDie** [comment]
@RyanCavanaugh Understood, but I still think the error message is misleading, since it mentions only half of what's going on (it doesn't say that a parameter is missing).

**Comment by RyanCavanaugh** [comment]
We could show the errors in *both* directions, but that would usually imply to the user that there are twice as many problems as there actually are.

**Comment by LuxDie** [comment]
Maybe don't comment on the missing property? Just say "neither type is assignable to each other"?

**Comment by RyanCavanaugh** [comment]
With no elaboration?

**Comment by LuxDie** [comment]
Maybe "both types are missing a required element"?

**Comment by RyanCavanaugh** [comment]
The number of permutations here is effectively infinite; we can't come up with a different error message for every possible error case. What would we do with something like this?
```ts
interface A {
  x?: number;
  y: string;
}
interface B {
  x: number;
  y: number;
}
var a: A;
var b = <B>a;
```
"Can't convert because one type has a required member that's optional in another which would be fine except that there's another member that is `string` when the other is `number`" ? It's intractable.

**Comment by LuxDie** [comment]
In that case I would settle with no elaboration. Having only a partial explanation is confusing in my opinion and can lead to incorrect handling by less than expert developers.

**Comment by kitsonk** [comment]
Personally I like the an elaboration so you can potentially try to resolve it instead of scratching your head wondering why.
